### ISSUE TITLE
FTPDumper does not clean up old downloaded files when ARCHIVE is set to False

### ISSUE THREAD

**Issue by newgene** [biothings/biothings.api#173]
When [ARCHIVE](https://github.com/biothings/biothings.api/blob/181a36fc2d5f782bb3608ec032891b0eaa9e7e1d/biothings/hub/dataload/dumper.py#L46) is set to False, the downloaded files will be saved to the same folder (Data_Archive_root/latest). Looks like FTPDumper does not clean up the old downloads.

Here is an example from mychem hub: [PubChemDumper](https://github.com/biothings/mychem.info/blob/master/src/hub/dataload/sources/pubchem/pubchem_dump.py).

**Comment by zcqian** [comment]
If @namespacestd0 hasn't started working on this, I can work on this one.

**Comment by namespacestd0** [comment]
Please go ahead.

**Comment by zcqian** [comment]
So I checked the implementation, and right now it seems no dumper is actively removing anything (no calls to `os.remove` or `os.unlink`)

Perhaps most data sources just overwrite old files, but for pubchem the files have different names? I will manually do a dump and confirm.

**Comment by newgene** [comment]
In general, it may always be good to double check the `to_dump_list` v.s. the actual files in the folder after downloads are done. At least warn there are some discrenpencies.

**Comment by zcqian** [comment]
In this case `BaseDumper.to_dump` only contains the files that actually needs downloading, which means it does not include files that exist on the remote side but is not updated compared to the one existing locally. Usually it expects that the remote to only add or modify files, not delete them. The entire dumper class is designed to only dump data, instead of performing a one way synchronization (which is more complicated).

In the case for PubChem, they don't usually delete files either, except that in November of 2019, they changed how they organize files, resulting in deletion of some files.

My suggestion is that we don't do changes, and simply delete the old files manually, as such events should be rare occurrences.

However, two ways we can clean up the old downloads:
1. Modify `PubChemDumper. post_dump` to remove files that does not exist on the remove.
2. Modify `BaseDumper` to support deleting files (i.e. actually make it one way sync, relatively more work)

**Comment by newgene** [comment]
@zcqian Some pseudocode in my mind to implement a generic solution in `BaseDumper` class:

```
class BaseDumper:
    SYNC_DATA_FOLDER = False    # set to True when ARCHIVE is True

def __init__(self, ...):
    ...
    self.full_file_list = None
    ...

def sync_data_folder(self, full_file_list, data_folder):
      #handle sync on data_folder based on full_file_list

def post_dump(self, ...):
      if self.SYNC_DATA_FOLDER:
            if self.full_file_list:
                   self.sync_data_folder(self.full_file_list, data_folder)
            else:
                   error("self.full_file_list needs to be populated first.")
      ...
```

Then `PubChemDumper` just need to do this:

```
class PubChemDumper(FTPDumper):
      ARCHIVE = True

def create_todump_list(self, force=False):
      ...
      #set self.full_file_list here
      ...
```

How about this?

**Comment by zcqian** [comment]
The structure is less of an issue, the one you proposed have pitfalls anyways.

Are you sure we need this functionality? This is the point that I’m trying to make. For what it’s worth `FTPDumper` is working exactly as as expected.

**Comment by zcqian** [comment]
Closed via #186
